docs/lang/control-and-nondeterminism.md

```diff
@@ -0,0 +1,509 @@
+![diag2](./img/diag2.png)
+
+Notice that specification `coord` describes one infinite computation (and
+infinitely many finite computations that are prefixes the infinite
+computation).  Specification `coord2` describes three infinite computations.
+Specification `coord3` describes infinitely many infinite computations: At
+every step, `Next` may choose between `y' = x` or `y' = x+1`.
+
+Why are these specifications so different? The answer lies in non-determinism.
+Specification `coord` is completely deterministic: There is just one state that
+evaluates `Init` to `TRUE`, and every state is the first component of exactly
+one transition, as specified by `Next`. Specification `coord2` has
+non-determinism in the operator `Init`. Specification `coord3` has
+non-determinism in operator `Next`.
+
+**Discussion.**
+So far we have been talking about the intuition. If you like to know more about
+the logic behind TLA+ and the semantics of TLA+, check Chapter 16 of
+[[Specifying Systems]] and [[The Specification Language TLA+]].
+
+When we look at the operators like `Init` and `Next` in our examples, we can
+guess the states and transitions. If we could ask our logician friend to guess
+the states and transitions for us every time we read a TLA+ specification, that
+would be great. But this approach does not scale well.
+
+*Can we explain non-determinism to a computer?* It turns out that we can.
+TLC and Apalache work because their authors managed to write programs
+that are reasoning about
+non-determinism. Of course, this comes with constraints on the structure of the
+specifications. After all, people are much better at solving certain logical
+puzzles than computers, though people get bored much faster than computers. 
+
+To understand how TLC enumerates states, check Chapter 14 of [[Specifying
+Systems]]. In the rest of this document, we focus on treatment of
+non-determinism that is closer to Apalache.
+
+## Explaining non-determinism to computers
+
+To see how a program could evaluate a TLA+ expression, we need two more
+ingredients: partial states and oracles.
+
+**Partial states.** We introduce a special value `Null` that we will use to
+denote that a state variable has not been assigned a value yet. Note that
+`Null` is not a standard keyword in TLA+, but it is our special keyword that we
+use for evaluating TLA+ expressions.  A _partial state_ is a mapping that
+assigns to every state variable either a TLA+ value or the special value
+`Null`.
+
+**Evaluating deterministic expressions.** Consider the specification `coord`,
+which was given above.  By starting with the partial state `[x |-> Null, y |->
+Null]`, we can see how to automatically evaluate the body of the operator
+`Init`:
+
+```tla
+x = 0 /\ y = 0 
+```
+
+By following [semantics of conjunction](./booleans.md), we see that `/\` is
+evaluated from left-to-right. The left-hand side equality `x = 0` is treated
+as an assignment to `x`, since `x` has value `Null` in the partial state 
```

Comment:
It is not about primed variables, but about whether a variable (primed or not) is bound to a value. In Apalache, we use `x' = e` even in `Init` just because it is more convenient to process such expressions.